rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+with a higher severity are delivered first without a significant delay. The prioritization mechanism is based on
+priority property value: the notification with the higher value is processed earlier to provide an experience as close to real-time as possible.
+
+##### Notification storage
+
+NS storage keeps different notification statuses for every user. Most Kibana deployments have [from 20 to 90 users](https://elastic.slack.com/archives/C9097ABGC/p1597162974399900),
+although there are some odd instances with >100k users. So we consider two options at the moment (see the [workflow](https://docs.google.com/document/d/18s-BTHog_oPaQKf871gzuhxkah4pK1GYJM9DCLceaLo/edit#)):
+- Store notification state separately from a notification. State for new notifications created for every user when they change the state.
+```typescript
+interface  NotificationState {
+  notification_id: string;
+  recipient_id: string;
+  read_status: boolean;
+  pinned_status: boolean;
+}
+``` 
+This option consumes less storage space but might lead to expensive read operations due to an additional JOIN call to 
+search for notification state objects. We don’t expect Kibana to read more than 10-20 notifications at once, so the overhead might be acceptable.
+- Create a copy for every notification in a user-specific list of notifications. It’s [the recommended way to store data](https://www.elastic.co/guide/en/elasticsearch/reference/current/joining-queries.html),
+but it leads to duplicating notification messages. 
+
+Stale notifications and their states must be removed from the storage. There should be a dedicated task performing
+look up through the storage to remove notifications with *expired_at* > the current timestamp.
+
+Data are stored in the hidden system index to prevent accidental access by a random user.
+We don’t expect notifications to contain sensitive information.
+They must be stored in an encrypted form otherwise. 
+Kibana can query user-specific notifications by *recipent_id*.
+
+Alternative implementation might be to use SO with OLS([#17888](https://github.com/elastic/kibana/issues/17888)),
+but it might be overhead as we don't need any advanced SO functionality like export/import.
+Also, SO search functionality is quite limited and doesn’t allow using complex search queries. 
+
+### RemoteRepository
+Cloud plugin registered RemoteRepository as a strategy Kibana Notification Service for creation, storage,
+and retrieval notifications from the remote Unified Notification Service. 
+It’s still not clear what functionality is supported by UNS, but we expect it to be:
+- Create a notification for a particular recipient, a group of recipients. 
+- Read a list notification for a particular recipient.
+- Filter notification list.
+- Change a notification state.
+
+Kibana-specific entities shouldn’t leak to UNS: 
+- Kibana sends message translations for all the supported locales.
```

Comment:
Can elaborate on what this means? We would send the fully translated string, one for each locale?

Reply:
yes. If we want to support i18n on Cloud and show notifications across deployments with different Kibana versions. @jowiho Can you see a better option? This one consumes a lot of disk space and network bandwidth, but a notification contains everything that's needed to show it in UI.

Reply:
I just commented above that we could store the notification in a single language, according to the user's preference at that time. Not sure what the best interface would be. Either the notification source would first lookup the user's language preference and then generate a notification in that language. Or the source would generate a notification in all languages, and the notification service would lookup op the preferred language and store the notification in that language.